### PR TITLE
Fixes #4229 Refactor rename incorrect when using import *

### LINKED ISSUE #4229: Refactor rename incorrect when using import *
app.py
```
from class1 import *
obj = oldname()
```

class1.py:
```
class oldname:
    """description"""
```

Select `oldname` from app.py and refactor->rename
In dialog, "new name" is `*` which is incorrect

image

Type a new name and hit enter

Assertion
```
---------------------------
Assertion Failed: Abort=Quit, Retry=Debug, Ignore=Continue
---------------------------
Failed to find '*' in 'obj = oldname()' because we had no private prefix



   at Microsoft.PythonTools.Refactoring.LocationPreviewItem..ctor(VsProjectAnalyzer analyzer, FilePreviewItem parent, LocationInfo locationInfo, VariableType type) in C:\Projects\GitHub\huguesv\PTVS\Python\Product\PythonTools\PythonTools\Refactoring\LocationPreviewItem.cs:line 68

   at Microsoft.PythonTools.Refactoring.PreviewChangesEngine.CreatePreviewItems() in C:\Projects\GitHub\huguesv\PTVS\Python\Product\PythonTools\PythonTools\Refactoring\PreviewChangesEngine.cs:line 74

   at Microsoft.PythonTools.Refactoring.PreviewChangesEngine..ctor(IServiceProvider serviceProvider, IRenameVariableInput input, String expr, RenameVariableRequest request, String originalName, String privatePrefix, VsProjectAnalyzer analyzer, IEnumerable`1 variables) in C:\Projects\GitHub\huguesv\PTVS\Python\Product\PythonTools\PythonTools\Refactoring\PreviewChangesEngine.cs:line 47

   at Microsoft.PythonTools.Refactoring.VariableRenamer.<RenameVariable>d__4.MoveNext() in C:\Projects\GitHub\huguesv\PTVS\Python\Product\PythonTools\PythonTools\Refactoring\VariableRenamer.cs:line 111

   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.InvokeMoveNext(Object stateMachine)

   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)

   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)

   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.Run()

   at System.Threading.Tasks.SynchronizationContextAwaitTaskContinuation.<>c.<.cctor>b__8_0(Object state)

   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)

   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)

   at System.Windows.Threading.DispatcherOperation.InvokeImpl()

   at System.Windows.Threading.DispatcherOperation.InvokeInSecurityContext(Object state)

   at MS.Internal.CulturePreservingExecutionContext.CallbackWrapper(Object obj)

   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)

   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)

   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)

   at MS.Internal.CulturePreservingExecutionContext.Run(CulturePreservingExecutionContext executionContext, ContextCallback callback, Object state)

   at System.Windows.Threading.DispatcherOperation.Invoke()

   at System.Windows.Threading.Dispatcher.ProcessQueue()

   at System.Windows.Threading.Dispatcher.WndProcHook(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)

   at MS.Win32.HwndWrapper.WndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)

   at MS.Win32.HwndSubclass.DispatcherCallbackOperation(Object o)

   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)

   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)

   at System.Windows.Threading.Dispatcher.LegacyInvokeImpl(DispatcherPriority priority, TimeSpan timeout, Delegate method, Object args, Int32 numArgs)

   at MS.Win32.HwndSubclass.SubclassWndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam)


---------------------------
Abort   Retry   Ignore   
---------------------------
```

Then the result is this:
image

### PR DESCRIPTION
**Pull request by zooba**
Fixes #4229 Refactor rename incorrect when using import *
Use correct references when importing modules or members
Streamlines import analysis code
Fixes dependencies not being notified of new module members
Fixes transitive module dependencies
Fixes analysis log not recording all messages
Fixes some handling of private name prefixes

### CODE REVIEW

**Comment by huguesv** [comment]
`NavigableTests.ModuleDefinition` and `NavigableTests.ModuleImportDefinition` are failing so I did some manual testing and noticed that go to definition now has different behavior when you press F12 on a module name. Rather than give you multiple results for the import statement and the module file, now it just goes to the module file. I think that's better. We've had user complain of the old behavior.

I just want to make sure this is intentional. Also those failing tests should be updated.

**Comment by huguesv** [comment]
F12 on imported functions have similar new behavior.
```
from os import chdir

chdir()
```

F12 on the call, and it goes directly to the implementation. Again, I think that's best. Find all references does include the import statement in its results.

File: Python/Product/Analysis/Analyzer/DDG.cs [hunk 1]
```diff
@@ -252,53 +252,108 @@ internal class DDG : PythonWalker {
...
+            if (module?.Module == null) {
+                return null;
+            }
+
+            if (attribute == null || attribute.Count == 0) {
+                return module.AnalysisModule;
+            }
+
+            var value = module.Module.GetModuleMember(node, _unit, attribute[0], addRef, Scope, attribute.Count == 1 ? linkToName : null);
+            foreach (var n in attribute.Skip(1)) {
```
Comment (AlexanderSher) [inline Python/Product/Analysis/Analyzer/DDG.cs]: What attribute is skipped here?

File: Python/Product/Analysis/Analyzer/DDG.cs [hunk 2]
```diff
@@ -252,53 +252,108 @@ internal class DDG : PythonWalker {
...
+        /// </param>
+        /// <param name="node">
+        /// If <paramref name="addRef"/> is true, this node will be
+        /// added as a reference of the imported value.
+        /// </param>
+        /// <param name="addRef">
+        /// True to add <paramref name="node"/> as a reference of the
+        /// imported value.
+        /// </param>
+        private void ImportModuleOrMember(ModuleReference module, IReadOnlyList<string> attribute, string importAs, Node node, bool addRef) {
```
Comment (AlexanderSher) [inline Python/Product/Analysis/Analyzer/DDG.cs]: Can we call it `FinishImportModuleOrMember`?

File: Python/Product/Analysis/Analyzer/DDG.cs (hunk 1 above)
Comment (zooba) [reply Python/Product/Analysis/Analyzer/DDG.cs]: The one from the module. Say we are getting `re.RegexFlag.IGNORECASE`. That would result in us importing `re` and attribute would be `['RegexFlag', 'IGNORECASE']`.

We want to get `RegexFlag` directly from the module using `GetModuleMember`, and then use the generalized `GetMember` to get `IGNORECASE` from whatever is returned. So we skip `attribute[0]`.

**Pull request merged**